### PR TITLE
Lazily initialize ModuleInfo.BuildId

### PR DESCRIPTION
**Pull request by NextTurn**
Contributes to #625 

Some questions:

- `IDataReader.GetVersionInfo` is a `void` method with an `out` parameter?

### CODE REVIEW

File: src/Microsoft.Diagnostics.Runtime/src/DataReaders/Core/CoreDumpReader.cs [hunk 1]
```diff
@@ -141,6 +142,13 @@ public bool GetThreadContext(uint threadID, uint contextFlags, Span<byte> contex
             return false;
         }
 
+        public ImmutableArray<byte> GetBuildId(ulong baseAddress)
+        {
+            ElfLoadedImage image = _core.LoadedImages.First(image => (ulong)image.BaseAddress == baseAddress);
+            ElfFile? file = image.Open();
+            return file?.BuildId ?? default;
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/DataReaders/Core/CoreDumpReader.cs:149]:
```suggestion
            return file?.BuildId ?? ImmutableArray<byte>.Empty;
```

File: src/Microsoft.Diagnostics.Runtime/src/Linux/LinuxLiveDataReader.cs [hunk 2]
```diff
@@ -113,7 +114,7 @@ public void FlushCachedData()
             let containsExecutable = image.Any(entry => entry.IsExecutable)
             let beginAddress = image.Min(entry => entry.BeginAddress)
             let props = GetPEImageProperties(filePath)
-            select new ModuleInfo(beginAddress, props.Filesize, props.Timestamp, filePath, containsExecutable, buildId: default, version: props.Version);
+            select new ModuleInfo(beginAddress, props.Filesize, props.Timestamp, filePath, containsExecutable, buildId: ImmutableArray<byte>.Empty, version: props.Version);
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/Linux/LinuxLiveDataReader.cs:117]: Is build ID unavailable in live processes?

File: src/Microsoft.Diagnostics.Runtime/src/DataReaders/DbgEng/DbgEngDataReader.cs [hunk 3]
```diff
@@ -219,7 +220,7 @@ public IEnumerable<ModuleInfo> EnumerateModules()
                 for (int i = 0; i < bases.Length; ++i)
                 {
                     string? fn = _symbols.GetModuleNameStringWide(DebugModuleName.Image, i, bases[i]);
-                    ModuleInfo info = new ModuleInfo(bases[i], mods[i].Size, mods[i].TimeDateStamp, fn, isVirtual: true);
+                    ModuleInfo info = new ModuleInfo(bases[i], mods[i].Size, mods[i].TimeDateStamp, fn, isVirtual: true, buildId: ImmutableArray<byte>.Empty);
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/DataReaders/DbgEng/DbgEngDataReader.cs:223]: Maybe we should just remove the buildId parameter and always lazily calculate it?

File: src/Microsoft.Diagnostics.Runtime/src/Linux/LinuxLiveDataReader.cs (hunk 2 above)
Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/Linux/LinuxLiveDataReader.cs:117]: That original code is wrong.  It exists but it isn't needed by ClrMD itself.  In the live scenario we should almost always have libmscordaccore.so on disk (unless the user published a build and/or deleted the file from disk) so we wouldn't need to hit the symbol server for it.  (I also realize ClrMD hasn't added symbol server support for buildid based files.)

In this case though, I think removing the parameter from the constructor and always lazily pulling the value is the right approach.

**Comment by leculver** [comment]
> * `IDataReader.GetVersionInfo` is a `void` method with an `out` parameter?

That's one of many things I want to change on IDataReader, but I also want to do them all as one checkin because I know at least one team already using ClrMD 2.0 with a custom datareader and I'd like to minimize impact.

Some other IDataReader problems:

1. Remove `IsFullMemoryAvailable`, `EnumerateAllThreads`, etc that aren't used.
2. Remove `QueryMemory` or put it on IMemoryReader.
3. `GetVersionInfo` should return a bool.
3. Consider changing `IDataReader` to an abstract base class, or make `DataTarget`'s constructor take an abstract base class that contains a data reader instead of the current design.  Essentially I'd like to be able to add "optional" properties and functions to data readers without it being a breaking change.  Since we support desktop CLR, I can't use interface default methods...

It's on my todo list for the API review phase.  I'm wrapping up minidumps now.

File: src/Microsoft.Diagnostics.Runtime/src/DataReaders/DbgEng/DbgEngDataReader.cs (hunk 3 above)
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/DataReaders/DbgEng/DbgEngDataReader.cs:223]: Why lazily calculate if the build ID is known to be `Empty` on Windows?

Comment (leculver) [reply src/Microsoft.Diagnostics.Runtime/src/DataReaders/DbgEng/DbgEngDataReader.cs:223]: Ah right.  =)

**Comment by leculver** [comment]
@NextTurn This checkin may have conflicts with https://github.com/microsoft/clrmd/pull/635.  Please mark this PR as ready when it's ready for merge.  I'll be happy to merge this one first and resolve the conflicts if this one is done before https://github.com/microsoft/clrmd/pull/635.

**Pull request merged**